horovod/mxnet/mpi_ops.h

```diff
@@ -32,6 +32,46 @@ typedef ::mxnet::NDArray NDArray;
 typedef ::mxnet::engine::CallbackOnComplete CallbackOnComplete;
 typedef std::shared_ptr<MXTemporaryBuffer<NDArray>> MXTempBufferShared;
 
+struct MpiOpsParam {
+  std::shared_ptr<NDArray*> input;
```

Comment:
Since you're passing around a pointer to this structure, do fields of the structure need to be pointers, or could they be underlying objects?

Reply:
These fields also need to be shared_ptr to prevent them from being destroyed prematurely. I tried not to wrap them with shared_ptr but hit segfault.

Reply:
That's odd.  What kind of segfault were you getting?  The only thing that can be destroyed is `std::string`, but I imagine it gets copied around as needed.

Reply:
Yes, you are right. I confused myself with another error. Will change it.